testjson/execution.go

```diff
@@ -538,15 +546,41 @@ func (e *Execution) Failed() []TestCase {
 	return failed
 }
 
-// FilterFailedUnique filters a slice of failed TestCases by removing root test
-// case that have failed subtests.
+// FilterFailedUnique filters a slice of failed TestCases to remove any parent
+// tests that have failed subtests. The parent test will always be run when
+// running any of its subtests.
 func FilterFailedUnique(tcs []TestCase) []TestCase {
-	var result []TestCase
+	sort.Slice(tcs, func(i, j int) bool {
+		a, b := tcs[i], tcs[j]
+		if a.Package != b.Package {
+			return a.Package < b.Package
+		}
+		return len(a.Test.Name()) > len(b.Test.Name())
```

Comment:
With this approach it might be possible to avoid both of these sorts if we iterate over the test cases in reverse order, and then build the results with prepend instead of append. But maybe that relies too much on an assumed order of the data. I need to think about that a bit more.

Reply:
I guess this won't work because the fail event of the parent test is later in the output. So we do need this sort.